### PR TITLE
Make focusing visualization possible #900

### PR DESCRIPTION
**Pull request by ljkania**
This pr introduces the correct implementation of visualization logic. Elements are created only when necessary and updates are kept to the minimum. However there is a few issues we need to handle:
- There is some styling needed (visualization toggler is simple square, menu does not work well because of some problems with hover, background is transparent but should have color of background)
- We need to implement additional modes which show visualization full screen. To do that we need to add another layer first - this one will not be scaled and will be also used by searcher.
- We need to replace iframes with shadow dom since now focusing requires to rerender iframe loosing its state.

Awesome things this introduces:
- Extra layer for dom elements on top of base 3 layers of basegl. This one is for dom elements that should be in front (like focused visualization) and should be scaled. This layer does not affect any mouse events for base mechanism unless there is an element in this layer under the mouse.
- Two base modes for visualization: default and focused. The first one makes visualization non-responding to events and covered with basegl shape (visualization is in `layer-dom` from basegl at this state). After click on that shape visualization is focused - it is moved to top layer and events are enabled. Now everything should be under this visualization. Any outside action (but `mouseenter`, `mouseleave`, etc.) make visualization unfocus.
- Toggling node value between short value or error and visualization.

### CODE REVIEW

File: src/shape/Node.coffee [hunk 1]
```diff
@@ -19,6 +19,16 @@ export width = nodeRadius * 2 + nodeSelectionBorderMaxSize * 2
 export height = nodeRadius * 2 + nodeSelectionBorderMaxSize * 2
 export slope = 20
 
+export togglerSize = 10 
+export foldVisualizationButtonShape = basegl.expr ->
+    shape = plane()
+    white = Color.rgb [1,1,1]
+    shape.fill white
+export unfoldVisualizationButtonShape = basegl.expr ->
```
Comment (kustosz) [inline src/shape/Node.coffee]: Why this duplication?

Comment (ljkania) [reply src/shape/Node.coffee]: Those should be two different shapes. Just my skills are not good enough and now i can only make circles and rectangles

File: src/view/NodeEditor.coffee [hunk 2]
```diff
@@ -22,6 +26,12 @@ export class NodeEditor extends Disposable
         @visualizerLibraries ?= {}
         @inTransaction        = false
         @pending              = []
+        @topDomScene          = @_scene.addDOMScene()
```
Comment (pmlodawski) [inline src/view/NodeEditor.coffee]: move all scene modifications to `install` function from `main.coffee`

File: src/view/Visualization.coffee [hunk 3]
```diff
@@ -149,122 +131,176 @@ export class VisualizationContainer extends Widget
...
-    constructor: (args...) ->
-        super args...
+    cons: (values, @parent) =>
+        @nodeEditor = @parent.nodeEditor
+        @eventPath  = @parent.eventPath
+        @def        = menuToggler
         @configure
-            height: shape.height
-            width:  shape.width
+            height: @parent.height
```
Comment (pmlodawski) [inline src/view/Visualization.coffee]: set `minHeight`, `maxHeight`, `minWidth`, `maxWidth` instead

File: src/view/Visualization.coffee [hunk 4]
```diff
@@ -149,122 +131,176 @@ export class VisualizationContainer extends Widget
...
+        @view?.domElement.id = @key
+        @updateMenu()
 
+    updateView: =>
+        @group?.position.xy = [@position[0] - @width/2 , @position[1]]
 
     pushSelectVisualizer: (visId) => 
-        path = @eventPathh()
+        path = @eventPath()
         base = {
```
Comment (pmlodawski) [inline src/view/Visualization.coffee]: braces are not required

**Pull request merged**